### PR TITLE
Maintain stable ordering for P2P shuffling

### LINKED ISSUE #8421: DataFrame P2P Shuffling with stable ordering
The P2P algorithm as is currently does not strictly guarantee ordering. This can be problematic for some order sensitive operations like groupby + first (https://github.com/dask/dask/issues/10034) or for a `drop_duplicates` with `keep` (https://github.com/dask/dask/issues/10708)

It's a little work but should be possible to get P2P to be stable

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #8421

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by hendrikmakait** [comment]
## A/B test results

### TL;DR

`test_q6` shows a significant increase in peak (and avg.) memory. It's hard to tell how bad this is in reality given the ~10s duration of the query. I'm inclined to move forward with this PR.

### Runtime
Screenshot 2024-01-12 at 09 44 38

This looks like noise, `test_set_index_on_uber_lyft` got messed up by spot interruption.

### Average Memory
Screenshot 2024-01-12 at 09 45 03

Mostly noise, `test_q6` looks significant (more on that later).

### Peak Memory

Screenshot 2024-01-12 at 09 45 11

Once again noise with `test_q6` standing out.


### test_q6

Screenshot 2024-01-12 at 09 45 50
Screenshot 2024-01-12 at 09 45 56

Looking at the memory plots, it seems as though this PR has a significant impact on peak memory usage. Average memory also increases but remains pretty small.

**Review by fjetter — APPROVED** [review]

**Comment by crusaderky** [comment]
> Looking at the memory plots, it seems as though this PR has a significant impact on peak memory usage. Average memory also increases but remains pretty small.

I'm not particularly worried about +300 MiB per worker peak memory usage.
Particularly as the baseline is very very low to begin with.
